File: Python/130. SurroundedRegions.py
class Solution:
    def dfs(self, i, j):
        if i<0 or j<0 or i>=self.M or j>=self.N or self.board[i][j] != "O":
            return
        self.board[i][j] = 'T'
        neib_list = [[i+1,j],[i-1,j],[i,j-1],[i,j+1]]
        for x, y in neib_list:
            self.dfs(x, y)
    
    def solve(self, board):
        if not board: return 0
        self.board, self.M, self.N = board, len(board), len(board[0])
        
        for i in range(0, self.M):
            self.dfs(i,0)
            self.dfs(i,self.N-1)
            
        for j in range(0, self.N):
            self.dfs(0,j)
            self.dfs(self.M-1,j)
        
        for i,j in product(range(self.M), range(self.N)):
            board[i][j] = "X" if board[i][j] != "T" else "O"

File: C++/130. Surrounded Regions.cpp
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        if (board.size() == 0 || board[0].size() == 0)
            return;
        
        //left border
        for (int row = 0;  row < board.size(); ++row)
            if (board[row][0] == 'O')
                floodFill(board, row, 0, 'O', 'T');
        //top border
        for (int col = 1; col < board[0].size(); ++col)
            if (board[0][col] == 'O')
                floodFill(board, 0, col, 'O', 'T');      
        //right border
        for (int row = 1; row < board.size(); ++row)
            if (board[row][board[0].size() - 1] == 'O')
                floodFill(board, row, board[0].size() - 1, 'O', 'T'); 
        //bottom border
        for (int col = 1; col < board[0].size() - 1; ++col)
            if (board[board.size() - 1][col] == 'O')  
                floodFill(board, board.size() - 1, col, 'O', 'T');   
        
        for (int row = 0; row < board.size(); ++row)
            for (int col = 0; col < board[0].size(); ++col)
            {
                if (board[row][col] == 'T')
                    board[row][col] = 'O';
                else
                    board[row][col] = 'X';
            }
    }
private:
    void floodFill(vector<vector<char>>& board, int row, int col, char origCh, char newCh) {
        if (row < 0 || row >= board.size() || col < 0 || col >= board[0].size())
            return;
        if (board[row][col] != origCh)
            return;
        board[row][col] = newCh;
        floodFill(board, row, col - 1, origCh, newCh);
        floodFill(board, row - 1, col, origCh, newCh);
        floodFill(board, row, col + 1, origCh, newCh);
        floodFill(board, row + 1, col, origCh, newCh);
    }; 
};
